{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm",
      "gpuType": "A100"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "import pandas as pd\n",
        "\n",
        "FILE_PATH = \"/content/drive/MyDrive/genesExpression/RABC13_Level3.txt\"  # <-- change this\n",
        "\n",
        "df = pd.read_csv(FILE_PATH, sep=\"\\t\")\n",
        "print(df.shape)\n",
        "df.head()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 290
        },
        "id": "_sCVcEh3AGee",
        "outputId": "078290af-4738-44bb-c4f4-c4d61d81f1ab"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n",
            "(19921, 100)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  symbol  GSM1381406  GSM1381407  GSM1381408  GSM1381430  GSM1381431  \\\n",
              "0   A1BG    5.034736    5.153410    4.771349    5.489528    4.913313   \n",
              "1   A1CF    3.580576    4.112098    4.133986    4.167102    3.732367   \n",
              "2  A2BP1    2.817091    2.817612    3.048173    3.198213    2.944027   \n",
              "3  A2LD1    4.053511    4.305152    4.077550    4.042980    3.934547   \n",
              "4    A2M    8.787091    9.036652    9.014434    8.947403    8.691395   \n",
              "\n",
              "   GSM1381432  GSM1381433  GSM1381524  GSM1381525  ...  GSM1382141  \\\n",
              "0    4.938768    5.051148    5.952392    6.285110  ...    6.155497   \n",
              "1    3.768247    4.379803    4.356622    4.453563  ...    4.203970   \n",
              "2    3.059729    2.869777    3.126752    3.136976  ...    3.100237   \n",
              "3    4.055266    3.759129    4.341665    4.666297  ...    4.232469   \n",
              "4    8.679810    8.834901    5.024601    5.072084  ...    4.659070   \n",
              "\n",
              "   GSM1382142  GSM1382143  GSM1382144  GSM1382145  GSM1382146  GSM1382147  \\\n",
              "0    6.130924    5.382765    5.785689    6.058182    5.783485    6.024936   \n",
              "1    4.258301    4.181870    4.213459    4.166540    4.122978    4.474931   \n",
              "2    2.988933    2.975656    3.886131    3.133297    3.045693    3.025082   \n",
              "3    4.259745    3.880452    4.016038    4.091444    4.148738    4.416004   \n",
              "4    5.100233    4.721538    4.840841    4.933884    4.634589    4.783980   \n",
              "\n",
              "   GSM1382148  GSM1382149  GSM1382150  \n",
              "0    5.708449    5.829147    5.930300  \n",
              "1    4.077597    4.103557    4.239628  \n",
              "2    3.007035    3.162358    3.071836  \n",
              "3    4.098573    4.411988    4.347975  \n",
              "4    4.863360    4.596491    4.830779  \n",
              "\n",
              "[5 rows x 100 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-7d9b1b6c-f197-46c8-b1e2-4d9b5a6a6ded\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>symbol</th>\n",
              "      <th>GSM1381406</th>\n",
              "      <th>GSM1381407</th>\n",
              "      <th>GSM1381408</th>\n",
              "      <th>GSM1381430</th>\n",
              "      <th>GSM1381431</th>\n",
              "      <th>GSM1381432</th>\n",
              "      <th>GSM1381433</th>\n",
              "      <th>GSM1381524</th>\n",
              "      <th>GSM1381525</th>\n",
              "      <th>...</th>\n",
              "      <th>GSM1382141</th>\n",
              "      <th>GSM1382142</th>\n",
              "      <th>GSM1382143</th>\n",
              "      <th>GSM1382144</th>\n",
              "      <th>GSM1382145</th>\n",
              "      <th>GSM1382146</th>\n",
              "      <th>GSM1382147</th>\n",
              "      <th>GSM1382148</th>\n",
              "      <th>GSM1382149</th>\n",
              "      <th>GSM1382150</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>A1BG</td>\n",
              "      <td>5.034736</td>\n",
              "      <td>5.153410</td>\n",
              "      <td>4.771349</td>\n",
              "      <td>5.489528</td>\n",
              "      <td>4.913313</td>\n",
              "      <td>4.938768</td>\n",
              "      <td>5.051148</td>\n",
              "      <td>5.952392</td>\n",
              "      <td>6.285110</td>\n",
              "      <td>...</td>\n",
              "      <td>6.155497</td>\n",
              "      <td>6.130924</td>\n",
              "      <td>5.382765</td>\n",
              "      <td>5.785689</td>\n",
              "      <td>6.058182</td>\n",
              "      <td>5.783485</td>\n",
              "      <td>6.024936</td>\n",
              "      <td>5.708449</td>\n",
              "      <td>5.829147</td>\n",
              "      <td>5.930300</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>A1CF</td>\n",
              "      <td>3.580576</td>\n",
              "      <td>4.112098</td>\n",
              "      <td>4.133986</td>\n",
              "      <td>4.167102</td>\n",
              "      <td>3.732367</td>\n",
              "      <td>3.768247</td>\n",
              "      <td>4.379803</td>\n",
              "      <td>4.356622</td>\n",
              "      <td>4.453563</td>\n",
              "      <td>...</td>\n",
              "      <td>4.203970</td>\n",
              "      <td>4.258301</td>\n",
              "      <td>4.181870</td>\n",
              "      <td>4.213459</td>\n",
              "      <td>4.166540</td>\n",
              "      <td>4.122978</td>\n",
              "      <td>4.474931</td>\n",
              "      <td>4.077597</td>\n",
              "      <td>4.103557</td>\n",
              "      <td>4.239628</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>A2BP1</td>\n",
              "      <td>2.817091</td>\n",
              "      <td>2.817612</td>\n",
              "      <td>3.048173</td>\n",
              "      <td>3.198213</td>\n",
              "      <td>2.944027</td>\n",
              "      <td>3.059729</td>\n",
              "      <td>2.869777</td>\n",
              "      <td>3.126752</td>\n",
              "      <td>3.136976</td>\n",
              "      <td>...</td>\n",
              "      <td>3.100237</td>\n",
              "      <td>2.988933</td>\n",
              "      <td>2.975656</td>\n",
              "      <td>3.886131</td>\n",
              "      <td>3.133297</td>\n",
              "      <td>3.045693</td>\n",
              "      <td>3.025082</td>\n",
              "      <td>3.007035</td>\n",
              "      <td>3.162358</td>\n",
              "      <td>3.071836</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>A2LD1</td>\n",
              "      <td>4.053511</td>\n",
              "      <td>4.305152</td>\n",
              "      <td>4.077550</td>\n",
              "      <td>4.042980</td>\n",
              "      <td>3.934547</td>\n",
              "      <td>4.055266</td>\n",
              "      <td>3.759129</td>\n",
              "      <td>4.341665</td>\n",
              "      <td>4.666297</td>\n",
              "      <td>...</td>\n",
              "      <td>4.232469</td>\n",
              "      <td>4.259745</td>\n",
              "      <td>3.880452</td>\n",
              "      <td>4.016038</td>\n",
              "      <td>4.091444</td>\n",
              "      <td>4.148738</td>\n",
              "      <td>4.416004</td>\n",
              "      <td>4.098573</td>\n",
              "      <td>4.411988</td>\n",
              "      <td>4.347975</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>A2M</td>\n",
              "      <td>8.787091</td>\n",
              "      <td>9.036652</td>\n",
              "      <td>9.014434</td>\n",
              "      <td>8.947403</td>\n",
              "      <td>8.691395</td>\n",
              "      <td>8.679810</td>\n",
              "      <td>8.834901</td>\n",
              "      <td>5.024601</td>\n",
              "      <td>5.072084</td>\n",
              "      <td>...</td>\n",
              "      <td>4.659070</td>\n",
              "      <td>5.100233</td>\n",
              "      <td>4.721538</td>\n",
              "      <td>4.840841</td>\n",
              "      <td>4.933884</td>\n",
              "      <td>4.634589</td>\n",
              "      <td>4.783980</td>\n",
              "      <td>4.863360</td>\n",
              "      <td>4.596491</td>\n",
              "      <td>4.830779</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 100 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7d9b1b6c-f197-46c8-b1e2-4d9b5a6a6ded')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-7d9b1b6c-f197-46c8-b1e2-4d9b5a6a6ded button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-7d9b1b6c-f197-46c8-b1e2-4d9b5a6a6ded');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 1
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "\n",
        "# Try common gene column names\n",
        "possible_gene_cols = [\"symbol\", \"gene\", \"Gene\", \"GENE\", \"gene_symbol\"]\n",
        "gene_col = next((c for c in possible_gene_cols if c in df.columns), df.columns[0])\n",
        "\n",
        "# Drop obvious non-expression columns (optional)\n",
        "drop_cols = [c for c in [\"index\", \"Index\", \"ID\", \"id\"] if c in df.columns]\n",
        "\n",
        "work = df.drop(columns=drop_cols, errors=\"ignore\").copy()\n",
        "\n",
        "# Gene column\n",
        "work[gene_col] = work[gene_col].astype(str)\n",
        "\n",
        "# Keep numeric expression columns only (these are your samples)\n",
        "numeric_cols = []\n",
        "for c in work.columns:\n",
        "    if c == gene_col:\n",
        "        continue\n",
        "    s = pd.to_numeric(work[c], errors=\"coerce\")\n",
        "    if s.notna().mean() > 0.8:  # at least 80% numeric\n",
        "        numeric_cols.append(c)\n",
        "\n",
        "print(\"Gene column:\", gene_col)\n",
        "print(\"Number of sample columns detected:\", len(numeric_cols))\n",
        "print(\"First sample columns:\", numeric_cols[:5])\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0m2Jd30ZAjOi",
        "outputId": "309e09a3-3711-4b9d-b452-532633905f01"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Gene column: symbol\n",
            "Number of sample columns detected: 99\n",
            "First sample columns: ['GSM1381406', 'GSM1381407', 'GSM1381408', 'GSM1381430', 'GSM1381431']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "3) Select one sample (patient) + compute up/down signature\n",
        "\n",
        "This is your “patient state” built from a cohort-style matrix."
      ],
      "metadata": {
        "id": "BbsOYN0-Anvp"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "B) Build a “patient signature” from expression (simple, works for demos)\n",
        "\n",
        "Pick one sample column (a GSM column) and compute z-scores across samples per gene:"
      ],
      "metadata": {
        "id": "98Ur-nmk_mnJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "expr = work[[gene_col] + numeric_cols].copy()\n",
        "expr = expr.dropna(subset=[gene_col]).drop_duplicates(subset=[gene_col])\n",
        "\n",
        "# Convert to numeric\n",
        "mat = expr[numeric_cols].apply(pd.to_numeric, errors=\"coerce\")\n",
        "mat.index = expr[gene_col].values\n",
        "#standarize format : DGIdb expects gene symbols like STAT3, not STAT3.1 or extra spaces.\n",
        "mat.index = (\n",
        "    mat.index.astype(str)\n",
        "    .str.upper()\n",
        "    .str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
        "    .str.strip()\n",
        ")\n",
        "\n",
        "# Pick a patient sample (change this to any column you want)\n",
        "patient_sample = numeric_cols[0]\n",
        "patient_vec = mat[patient_sample]\n",
        "\n",
        "# Z-score patient vs cohort for each gene\n",
        "mu = mat.mean(axis=1)\n",
        "sd = mat.std(axis=1).replace(0, np.nan)\n",
        "z = ((patient_vec - mu) / sd).dropna()\n",
        "\n",
        "up_genes = z.sort_values(ascending=False).head(200).index.tolist()\n",
        "down_genes = z.sort_values().head(200).index.tolist()\n",
        "\n",
        "print(\"Patient sample:\", patient_sample)\n",
        "print(\"Top up genes:\", up_genes[:10])\n",
        "print(\"Top down genes:\", down_genes[:10])\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0AqLIUaiAofC",
        "outputId": "089303c9-321a-4ec1-9f5e-443458272048",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Patient sample: GSM1381406\n",
            "Top up genes: ['LBP', 'SEMA3A', 'BAI3', 'RSPO2', 'IP6K3', 'WDR69', 'PAQR5', 'LRRC2', 'CCBE1', 'TRIM9']\n",
            "Top down genes: ['SUB1', 'RPL29', 'MATR3', 'HNRNPUL1', 'C2ORF24', 'EIF1', 'SHOC2', 'RPL30', 'KIF24', 'CNN2']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "C) DGIdb: map patient genes → candidate drugs\n",
        "\n",
        "Use dgipy, then inspect available query functions (so you don’t get stuck on function names):"
      ],
      "metadata": {
        "id": "pmO_onUR_spP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip -q uninstall -y gql dgipy graphql-core\n",
        "!pip -q install dgipy==0.1.0 \"gql[requests]==3.5.0\"\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_Sy8VRq0GMMf",
        "outputId": "4b4d7ca5-8b80-41aa-f9b6-0ac43a8c63cf",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33mWARNING: Skipping gql as it is not installed.\u001b[0m\u001b[33m\n",
            "\u001b[0m\u001b[33mWARNING: Skipping dgipy as it is not installed.\u001b[0m\u001b[33m\n",
            "\u001b[0m\u001b[33mWARNING: Skipping graphql-core as it is not installed.\u001b[0m\u001b[33m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.0/4.0 MB\u001b[0m \u001b[31m107.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m74.0/74.0 kB\u001b[0m \u001b[31m7.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m207.3/207.3 kB\u001b[0m \u001b[31m21.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.9/7.9 MB\u001b[0m \u001b[31m165.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m204.0/204.0 kB\u001b[0m \u001b[31m25.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Building wheel for dash_cytoscape (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "import dgipy\n",
        "[f for f in dir(dgipy) if f.startswith(\"get_\")]\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Q0xY_xeA_wp7",
        "outputId": "ab708bf6-0b01-42ec-b486-aa371760fd45"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['get_all_genes',\n",
              " 'get_categories',\n",
              " 'get_drug_applications',\n",
              " 'get_drugs',\n",
              " 'get_genes',\n",
              " 'get_interactions',\n",
              " 'get_sources']"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import inspect\n",
        "\n",
        "print(\"get_interactions signature:\", inspect.signature(dgipy.get_interactions))\n",
        "help(dgipy.get_interactions)\n"
      ],
      "metadata": {
        "id": "vnivjFyECjJ-",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "97b63303-54fc-449a-873e-97c124d66cfc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "get_interactions signature: (terms: list, search: str = 'genes', immunotherapy: bool | None = None, antineoplastic: bool | None = None, source: str | None = None, pmid: int | None = None, interaction_type: str | None = None, approved: str | None = None, api_url: str | None = None) -> dict\n",
            "Help on function get_interactions in module dgipy.dgidb:\n",
            "\n",
            "get_interactions(terms: list, search: str = 'genes', immunotherapy: bool | None = None, antineoplastic: bool | None = None, source: str | None = None, pmid: int | None = None, interaction_type: str | None = None, approved: str | None = None, api_url: str | None = None) -> dict\n",
            "    Perform an interaction look up for drugs or genes of interest\n",
            "\n",
            "    :param terms: drugs or genes for interaction look up\n",
            "    :param search: interaction search type. valid types are \"drugs\" or \"genes\"\n",
            "    :param immunotherapy: filter option for results that are used in immunotherapy\n",
            "    :param antineoplastic: filter option for results that are part of antineoplastic regimens\n",
            "    :param source: filter option for specific database of interest\n",
            "    :param pmid: filter option for specific PMID\n",
            "    :param interaction_type: filter option for specific interaction types\n",
            "    :param approved: filter option for approved interactions\n",
            "    :param api_url: API endpoint for GraphQL request\n",
            "    :return: interaction results for terms\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dgipy, gql\n",
        "print(\"dgipy:\", getattr(dgipy, \"__version__\", \"unknown\"))\n",
        "print(\"gql:\", gql.__version__)\n"
      ],
      "metadata": {
        "id": "G4lhMsyeGUVN",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ba8406ce-489c-4c1b-dc54-0cc1072206ae"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "dgipy: unknown\n",
            "gql: 3.5.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "creates interactions (the DGIdb results dataframe)"
      ],
      "metadata": {
        "id": "nQOCRAl7mGVk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "def chunks(lst, n=50):\n",
        "    for i in range(0, len(lst), n):\n",
        "        yield lst[i:i+n]\n",
        "\n",
        "parts = []\n",
        "for glist in chunks(up_genes[:200], 50):\n",
        "    res = dgipy.get_interactions(terms=glist, search=\"genes\")\n",
        "    parts.append(pd.DataFrame(res))\n",
        "\n",
        "interactions = pd.concat(parts, ignore_index=True) if parts else pd.DataFrame()\n",
        "print(\"Rows:\", len(interactions))\n",
        "print(\"Columns:\", interactions.columns.tolist())\n",
        "interactions.head()\n"
      ],
      "metadata": {
        "id": "YipFYI9hD0dK",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 608
        },
        "outputId": "4d72bd47-b186-476a-a40e-2ba8f0be4a6c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Rows: 866\n",
            "Columns: ['gene_name', 'gene_concept_id', 'gene_long_name', 'drug_name', 'drug_concept_id', 'drug_approved', 'interaction_score', 'interaction_attributes', 'interaction_sources', 'interaction_pmids']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  gene_name gene_concept_id  \\\n",
              "0    GABRA4       hgnc:4078   \n",
              "1    GABRA4       hgnc:4078   \n",
              "2    GABRA4       hgnc:4078   \n",
              "3    GABRA4       hgnc:4078   \n",
              "4    GABRA4       hgnc:4078   \n",
              "\n",
              "                                      gene_long_name                drug_name  \\\n",
              "0  gamma-aminobutyric acid type A receptor subuni...                 CLOBAZAM   \n",
              "1  gamma-aminobutyric acid type A receptor subuni...  CLORAZEPATE DIPOTASSIUM   \n",
              "2  gamma-aminobutyric acid type A receptor subuni...               GANAXOLONE   \n",
              "3  gamma-aminobutyric acid type A receptor subuni...               CLONAZEPAM   \n",
              "4  gamma-aminobutyric acid type A receptor subuni...               BROMAZEPAM   \n",
              "\n",
              "  drug_concept_id  drug_approved  interaction_score  \\\n",
              "0     rxcui:21241           True           0.085767   \n",
              "1      rxcui:2607           True           0.039310   \n",
              "2   rxcui:2604689           True           0.069885   \n",
              "3      rxcui:2598           True           0.075475   \n",
              "4      rxcui:1749           True           0.059901   \n",
              "\n",
              "                              interaction_attributes  \\\n",
              "0  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "1  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "2  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "3  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "4  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "\n",
              "                interaction_sources interaction_pmids  \n",
              "0  [TEND, ChEMBL, TdgClinicalTrial]                []  \n",
              "1                          [ChEMBL]                []  \n",
              "2        [ChEMBL, TdgClinicalTrial]                []  \n",
              "3  [ChEMBL, TEND, TdgClinicalTrial]                []  \n",
              "4          [TdgClinicalTrial, TEND]                []  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0b6e887d-f034-4a51-879b-65a962727e05\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_name</th>\n",
              "      <th>gene_concept_id</th>\n",
              "      <th>gene_long_name</th>\n",
              "      <th>drug_name</th>\n",
              "      <th>drug_concept_id</th>\n",
              "      <th>drug_approved</th>\n",
              "      <th>interaction_score</th>\n",
              "      <th>interaction_attributes</th>\n",
              "      <th>interaction_sources</th>\n",
              "      <th>interaction_pmids</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>CLOBAZAM</td>\n",
              "      <td>rxcui:21241</td>\n",
              "      <td>True</td>\n",
              "      <td>0.085767</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[TEND, ChEMBL, TdgClinicalTrial]</td>\n",
              "      <td>[]</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>CLORAZEPATE DIPOTASSIUM</td>\n",
              "      <td>rxcui:2607</td>\n",
              "      <td>True</td>\n",
              "      <td>0.039310</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[ChEMBL]</td>\n",
              "      <td>[]</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>GANAXOLONE</td>\n",
              "      <td>rxcui:2604689</td>\n",
              "      <td>True</td>\n",
              "      <td>0.069885</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[ChEMBL, TdgClinicalTrial]</td>\n",
              "      <td>[]</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>CLONAZEPAM</td>\n",
              "      <td>rxcui:2598</td>\n",
              "      <td>True</td>\n",
              "      <td>0.075475</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[ChEMBL, TEND, TdgClinicalTrial]</td>\n",
              "      <td>[]</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>BROMAZEPAM</td>\n",
              "      <td>rxcui:1749</td>\n",
              "      <td>True</td>\n",
              "      <td>0.059901</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[TdgClinicalTrial, TEND]</td>\n",
              "      <td>[]</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0b6e887d-f034-4a51-879b-65a962727e05')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-0b6e887d-f034-4a51-879b-65a962727e05 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-0b6e887d-f034-4a51-879b-65a962727e05');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "interactions",
              "summary": "{\n  \"name\": \"interactions\",\n  \"rows\": 866,\n  \"fields\": [\n    {\n      \"column\": \"gene_name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 55,\n        \"samples\": [\n          \"HOXB6\",\n          \"EPCAM\",\n          \"GABRB2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gene_concept_id\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 55,\n        \"samples\": [\n          \"hgnc:5117\",\n          \"hgnc:11529\",\n          \"hgnc:4082\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gene_long_name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 55,\n        \"samples\": [\n          \"homeobox B6\",\n          \"epithelial cell adhesion molecule\",\n          \"gamma-aminobutyric acid type A receptor subunit beta2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"drug_name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 664,\n        \"samples\": [\n          \"AMINOGLUTETHIMIDE\",\n          \"CELECOXIB\",\n          \"TREPROSTINIL\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"drug_concept_id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 665,\n        \"samples\": [\n          \"rxcui:7004\",\n          \"chembl:CHEMBL67511\",\n          \"iuphar.ligand:12520\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"drug_approved\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          false,\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"interaction_score\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 4.0282612547557575,\n        \"min\": 0.006604731599049283,\n        \"max\": 52.20379855888553,\n        \"num_unique_values\": 376,\n        \"samples\": [\n          6.023515218332946,\n          0.1882348505729045\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"interaction_attributes\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"interaction_sources\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"interaction_pmids\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "since we are going to merge DGIdb output with the IC50 output, so we must standardize."
      ],
      "metadata": {
        "id": "S9kjd1HbmeWm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# ===== Normalize DGIdb interactions dataframe schema (run AFTER interactions is created) =====\n",
        "print(\"Before normalization:\", interactions.columns.tolist())\n",
        "\n",
        "# Try to standardize likely column names coming from DGIdb\n",
        "rename_map = {}\n",
        "for c in interactions.columns:\n",
        "    lc = c.lower()\n",
        "\n",
        "    if lc in [\"gene_name\", \"gene\", \"genesymbol\", \"gene_symbol\", \"geneSymbol\".lower()]:\n",
        "        rename_map[c] = \"gene\"\n",
        "    if lc in [\"drug_name\", \"drug\", \"drugname\", \"drug_name_normalized\"]:\n",
        "        rename_map[c] = \"drug\"\n",
        "    if lc in [\"interaction_types\", \"interaction_type\", \"interactiontype\", \"types\"]:\n",
        "        rename_map[c] = \"interaction_types\"\n",
        "    if lc in [\"sources\", \"source\"]:\n",
        "        rename_map[c] = \"sources\"\n",
        "\n",
        "interactions = interactions.rename(columns=rename_map)\n",
        "\n",
        "# Ensure required columns exist\n",
        "for col in [\"gene\", \"drug\", \"interaction_types\", \"sources\"]:\n",
        "    if col not in interactions.columns:\n",
        "        interactions[col] = None\n",
        "\n",
        "# Clean values\n",
        "interactions[\"gene\"] = interactions[\"gene\"].astype(str).str.upper().str.strip()\n",
        "interactions[\"drug\"] = interactions[\"drug\"].astype(str).str.strip()\n",
        "\n",
        "print(\"After normalization:\", interactions.columns.tolist())\n",
        "print(\"Interactions shape:\", interactions.shape)\n",
        "interactions.head()\n"
      ],
      "metadata": {
        "id": "RWHAHVmrl-CM",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 625
        },
        "outputId": "874bf4cf-ffbf-449f-bee6-48c40677770e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Before normalization: ['gene_name', 'gene_concept_id', 'gene_long_name', 'drug_name', 'drug_concept_id', 'drug_approved', 'interaction_score', 'interaction_attributes', 'interaction_sources', 'interaction_pmids']\n",
            "After normalization: ['gene', 'gene_concept_id', 'gene_long_name', 'drug', 'drug_concept_id', 'drug_approved', 'interaction_score', 'interaction_attributes', 'interaction_sources', 'interaction_pmids', 'interaction_types', 'sources']\n",
            "Interactions shape: (866, 12)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     gene gene_concept_id                                     gene_long_name  \\\n",
              "0  GABRA4       hgnc:4078  gamma-aminobutyric acid type A receptor subuni...   \n",
              "1  GABRA4       hgnc:4078  gamma-aminobutyric acid type A receptor subuni...   \n",
              "2  GABRA4       hgnc:4078  gamma-aminobutyric acid type A receptor subuni...   \n",
              "3  GABRA4       hgnc:4078  gamma-aminobutyric acid type A receptor subuni...   \n",
              "4  GABRA4       hgnc:4078  gamma-aminobutyric acid type A receptor subuni...   \n",
              "\n",
              "                      drug drug_concept_id  drug_approved  interaction_score  \\\n",
              "0                 CLOBAZAM     rxcui:21241           True           0.085767   \n",
              "1  CLORAZEPATE DIPOTASSIUM      rxcui:2607           True           0.039310   \n",
              "2               GANAXOLONE   rxcui:2604689           True           0.069885   \n",
              "3               CLONAZEPAM      rxcui:2598           True           0.075475   \n",
              "4               BROMAZEPAM      rxcui:1749           True           0.059901   \n",
              "\n",
              "                              interaction_attributes  \\\n",
              "0  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "1  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "2  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "3  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "4  {'Interaction Context': None, 'Endogenous Drug...   \n",
              "\n",
              "                interaction_sources interaction_pmids interaction_types  \\\n",
              "0  [TEND, ChEMBL, TdgClinicalTrial]                []              None   \n",
              "1                          [ChEMBL]                []              None   \n",
              "2        [ChEMBL, TdgClinicalTrial]                []              None   \n",
              "3  [ChEMBL, TEND, TdgClinicalTrial]                []              None   \n",
              "4          [TdgClinicalTrial, TEND]                []              None   \n",
              "\n",
              "  sources  \n",
              "0    None  \n",
              "1    None  \n",
              "2    None  \n",
              "3    None  \n",
              "4    None  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6e98f996-a2c2-4099-8a86-21bb497e05ea\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene</th>\n",
              "      <th>gene_concept_id</th>\n",
              "      <th>gene_long_name</th>\n",
              "      <th>drug</th>\n",
              "      <th>drug_concept_id</th>\n",
              "      <th>drug_approved</th>\n",
              "      <th>interaction_score</th>\n",
              "      <th>interaction_attributes</th>\n",
              "      <th>interaction_sources</th>\n",
              "      <th>interaction_pmids</th>\n",
              "      <th>interaction_types</th>\n",
              "      <th>sources</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>CLOBAZAM</td>\n",
              "      <td>rxcui:21241</td>\n",
              "      <td>True</td>\n",
              "      <td>0.085767</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[TEND, ChEMBL, TdgClinicalTrial]</td>\n",
              "      <td>[]</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>CLORAZEPATE DIPOTASSIUM</td>\n",
              "      <td>rxcui:2607</td>\n",
              "      <td>True</td>\n",
              "      <td>0.039310</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[ChEMBL]</td>\n",
              "      <td>[]</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>GANAXOLONE</td>\n",
              "      <td>rxcui:2604689</td>\n",
              "      <td>True</td>\n",
              "      <td>0.069885</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[ChEMBL, TdgClinicalTrial]</td>\n",
              "      <td>[]</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>CLONAZEPAM</td>\n",
              "      <td>rxcui:2598</td>\n",
              "      <td>True</td>\n",
              "      <td>0.075475</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[ChEMBL, TEND, TdgClinicalTrial]</td>\n",
              "      <td>[]</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>GABRA4</td>\n",
              "      <td>hgnc:4078</td>\n",
              "      <td>gamma-aminobutyric acid type A receptor subuni...</td>\n",
              "      <td>BROMAZEPAM</td>\n",
              "      <td>rxcui:1749</td>\n",
              "      <td>True</td>\n",
              "      <td>0.059901</td>\n",
              "      <td>{'Interaction Context': None, 'Endogenous Drug...</td>\n",
              "      <td>[TdgClinicalTrial, TEND]</td>\n",
              "      <td>[]</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6e98f996-a2c2-4099-8a86-21bb497e05ea')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6e98f996-a2c2-4099-8a86-21bb497e05ea button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6e98f996-a2c2-4099-8a86-21bb497e05ea');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "interactions",
              "repr_error": "Out of range float values are not JSON compliant: nan"
            }
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Try to find drug/gene columns in the returned dataframe\n",
        "possible_gene_cols = [\"gene_name\", \"gene\", \"geneSymbol\", \"gene_symbol\"]\n",
        "possible_drug_cols = [\"drug_name\", \"drug\", \"drugName\", \"drug_name_normalized\"]\n",
        "\n",
        "gene_c = next((c for c in possible_gene_cols if c in interactions.columns), None)\n",
        "drug_c = next((c for c in possible_drug_cols if c in interactions.columns), None)\n",
        "\n",
        "print(\"Detected gene column:\", gene_c)\n",
        "print(\"Detected drug column:\", drug_c)\n",
        "\n",
        "if gene_c and drug_c:\n",
        "    interactions[gene_c] = interactions[gene_c].astype(str).str.upper()\n",
        "    up_set = set([g.upper() for g in up_genes])\n",
        "\n",
        "    # keep only interactions that match your up genes\n",
        "    inter_up = interactions[interactions[gene_c].isin(up_set)].copy()\n",
        "\n",
        "    # score drugs by number of unique up genes they hit\n",
        "    drug_scores = (\n",
        "    interactions.dropna(subset=[\"drug\"])\n",
        "    .groupby(\"drug\")[\"gene\"]\n",
        "    .nunique()\n",
        "    .sort_values(ascending=False)\n",
        "    .reset_index(name=\"num_patient_genes_hit\")\n",
        "    )\n",
        "# 20 kefaya 3shan akeed msh h model l adwya aktar mn kedaaaaaaaaaaa\n",
        "    TOP_N = 20\n",
        "    top_drugs = drug_scores.head(TOP_N)[\"drug\"].tolist()\n",
        "\n",
        "    print(\"Top drugs:\", top_drugs[:10])\n",
        "    drug_scores.head(20)\n",
        "\n",
        "else:\n",
        "    print(\"Could not detect gene/drug columns. Check interactions.columns and set them manually.\")\n"
      ],
      "metadata": {
        "id": "PmTh04u1DNaw",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "81eb2834-444e-4551-c340-30b22a9c8b66"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Detected gene column: gene\n",
            "Detected drug column: drug\n",
            "Top drugs: ['ENFLURANE', 'HALOTHANE', 'ISOFLURANE', 'DESFLURANE', 'SEVOFLURANE', 'METHOXYFLURANE', 'LORAZEPAM', 'TOPIRAMATE', 'PROPOFOL', 'REMIMAZOLAM BESYLATE']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "recommendation list: drugs most connected to the patient’s dysregulated genes (based on DGIdb)."
      ],
      "metadata": {
        "id": "UVqH9KEbNAEu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "drug_scores.head(20)\n"
      ],
      "metadata": {
        "id": "enTA2hIkMzIh",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 676
        },
        "outputId": "56623d1a-9633-407c-e848-601a27e6a141"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                       drug  num_patient_genes_hit\n",
              "0                 ENFLURANE                      6\n",
              "1                 HALOTHANE                      5\n",
              "2                ISOFLURANE                      5\n",
              "3                DESFLURANE                      5\n",
              "4               SEVOFLURANE                      5\n",
              "5            METHOXYFLURANE                      4\n",
              "6                 LORAZEPAM                      4\n",
              "7                TOPIRAMATE                      4\n",
              "8                  PROPOFOL                      4\n",
              "9      REMIMAZOLAM BESYLATE                      3\n",
              "10                 DIAZEPAM                      3\n",
              "11               RESEQUINIL                      3\n",
              "12              REMIMAZOLAM                      3\n",
              "13                 QUAZEPAM                      3\n",
              "14                DARIGABAT                      3\n",
              "15      SECOBARBITAL SODIUM                      3\n",
              "16              CLOTIAZEPAM                      3\n",
              "17  CLORAZEPATE DIPOTASSIUM                      3\n",
              "18              CLORAZEPATE                      3\n",
              "19               CLONAZEPAM                      3"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3baae7bd-9523-477a-ab97-6ce0df54f7c1\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>drug</th>\n",
              "      <th>num_patient_genes_hit</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENFLURANE</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>HALOTHANE</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ISOFLURANE</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>DESFLURANE</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SEVOFLURANE</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>METHOXYFLURANE</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>LORAZEPAM</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>TOPIRAMATE</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>PROPOFOL</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>REMIMAZOLAM BESYLATE</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>DIAZEPAM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>RESEQUINIL</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>REMIMAZOLAM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>QUAZEPAM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>DARIGABAT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SECOBARBITAL SODIUM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>CLOTIAZEPAM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>CLORAZEPATE DIPOTASSIUM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>CLORAZEPATE</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>CLONAZEPAM</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3baae7bd-9523-477a-ab97-6ce0df54f7c1')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3baae7bd-9523-477a-ab97-6ce0df54f7c1 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3baae7bd-9523-477a-ab97-6ce0df54f7c1');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "drug_scores",
              "summary": "{\n  \"name\": \"drug_scores\",\n  \"rows\": 664,\n  \"fields\": [\n    {\n      \"column\": \"drug\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 664,\n        \"samples\": [\n          \"TRIFLUOPERAZINE HYDROCHLORIDE\",\n          \"TG11-77\",\n          \"2-FUROYL-LIGRLO-AMIDE\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"num_patient_genes_hit\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 1,\n        \"max\": 6,\n        \"num_unique_values\": 6,\n        \"samples\": [\n          6,\n          5,\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        ") Build the patient feature vector (from your matrix)"
      ],
      "metadata": {
        "id": "96vrfP9qjcMi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "\n",
        "# Patient feature vector: use z-scores (recommended so scale matches)\n",
        "patient_features = z.copy()  # index = gene symbol, value = z-score\n",
        "patient_features.name = \"patient_z\"\n",
        "\n",
        "# Keep only finite\n",
        "patient_features = patient_features.replace([np.inf, -np.inf], np.nan).dropna()\n",
        "print(\"Patient features:\", patient_features.shape)\n"
      ],
      "metadata": {
        "id": "SQrVvANojc4z",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "02034633-a1d2-46e5-dbe6-8de2d1bca701"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Patient features: (19921,)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestRegressor\n",
        "from sklearn.metrics import mean_absolute_error, r2_score\n",
        "# ==== PRISM inputs (upload these to /content or set paths to where you saved them) ====\n",
        "PRISM_DOSE_RESPONSE_PATH = \"/content/drive/MyDrive/prisma ds/secondary-screen-dose-response-curve-parameters.csv\"\n",
        "PRISM_CELL_LINE_INFO_PATH = \"/content/drive/MyDrive/prisma ds/secondary-screen-cell-line-info.csv\"\n",
        "PRISM_TREATMENT_INFO_PATH = \"/content/drive/MyDrive/prisma ds/secondary-screen-replicate-collapsed-treatment-info.csv\"\n",
        "\n",
        "\n",
        "\n",
        "print(\"PRISM dose-response file:\", PRISM_DOSE_RESPONSE_PATH)\n",
        "print(\"PRISM cell-line info file:\", PRISM_CELL_LINE_INFO_PATH)\n",
        "print(\"PRISM treatment info file:\", PRISM_TREATMENT_INFO_PATH)\n"
      ],
      "metadata": {
        "id": "cEyfWV4RkFG4",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0f8ac0f0-c978-4c3a-db09-e91ddb6e4e83"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PRISM dose-response file: /content/drive/MyDrive/prisma ds/secondary-screen-dose-response-curve-parameters.csv\n",
            "PRISM cell-line info file: /content/drive/MyDrive/prisma ds/secondary-screen-cell-line-info.csv\n",
            "PRISM treatment info file: /content/drive/MyDrive/prisma ds/secondary-screen-replicate-collapsed-treatment-info.csv\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# ===== Expression file(s) =====\n",
        "# If this is YOUR patient expression (single patient), keep it here:\n",
        "PATIENT_EXPR_PATH = \"/content/drive/MyDrive/genesExpression/RABC13_Level3.txt\"\n",
        "\n",
        "# If you later download cell-line expression for PRISM/DepMap, put it here:\n",
        "# CELL_LINE_EXPR_PATH = \"/content/Cell_line_RMA_proc_basalExp.txt\"\n",
        "CELL_LINE_EXPR_PATH = None"
      ],
      "metadata": {
        "id": "aU9q6o4yTrwG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ==== Load PRISM secondary screen dose-response parameters ====\n",
        "prism = pd.read_csv(PRISM_DOSE_RESPONSE_PATH)\n",
        "\n",
        "# clean column names\n",
        "prism.columns = [c.strip() for c in prism.columns]\n",
        "\n",
        "print(\"PRISM shape:\", prism.shape)\n",
        "print(\"PRISM columns (first 40):\", prism.columns[:40].tolist())\n",
        "\n",
        "# --- Helper: pick the first existing column from a list ---\n",
        "def pick_col(df, candidates):\n",
        "    for c in candidates:\n",
        "        if c in df.columns:\n",
        "            return c\n",
        "    return None\n",
        "\n",
        "# ---- Detect key columns (these names vary by PRISM release) ----\n",
        "cell_col = pick_col(prism, [\"depmap_id\", \"DepMap_ID\", \"cell_line\", \"cell_line_name\", \"cell_id\", \"broad_id\"])\n",
        "drug_col = pick_col(prism, [\"name\", \"compound_name\", \"drug\", \"pert_iname\", \"broad_compound_id\", \"compound_id\"])\n",
        "\n",
        "# target candidates (PRISM usually provides AUC / EC50 / IC50 / logIC50-like columns depending on file)\n",
        "log10_ic50_col = pick_col(prism, [\"log10_ic50\", \"logIC50\", \"log_ic50\", \"log10IC50\"])\n",
        "ic50_uM_col     = pick_col(prism, [\"ic50\", \"IC50\", \"ic50_um\", \"IC50_uM\", \"ec50\", \"EC50\"])\n",
        "auc_col         = pick_col(prism, [\"auc\", \"AUC\", \"aac\", \"AAC\"])\n",
        "\n",
        "print(\"Detected cell_col:\", cell_col)\n",
        "print(\"Detected drug_col:\", drug_col)\n",
        "print(\"Detected log10_ic50_col:\", log10_ic50_col)\n",
        "print(\"Detected ic50_uM_col:\", ic50_uM_col)\n",
        "print(\"Detected auc_col:\", auc_col)\n",
        "\n",
        "if cell_col is None or drug_col is None:\n",
        "    raise ValueError(\"Could not detect cell line column or drug column. Show me prism.columns and I will map them exactly.\")\n"
      ],
      "metadata": {
        "id": "AfOj8DhHN7FW",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6d6425b8-734d-45ca-90dd-90f01d4267be"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/tmp/ipython-input-2439368592.py:2: DtypeWarning: Columns (14,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
            "  prism = pd.read_csv(PRISM_DOSE_RESPONSE_PATH)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PRISM shape: (701004, 20)\n",
            "PRISM columns (first 40): ['broad_id', 'depmap_id', 'ccle_name', 'screen_id', 'upper_limit', 'lower_limit', 'slope', 'r2', 'auc', 'ec50', 'ic50', 'name', 'moa', 'target', 'disease.area', 'indication', 'smiles', 'phase', 'passed_str_profiling', 'row_name']\n",
            "Detected cell_col: depmap_id\n",
            "Detected drug_col: name\n",
            "Detected log10_ic50_col: None\n",
            "Detected ic50_uM_col: ic50\n",
            "Detected auc_col: auc\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Build the training target table (IC50 if exists, else AUC)"
      ],
      "metadata": {
        "id": "QSxGXvjTUT4w"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Build response table (cell line × drug × target)\n",
        "train_resp = prism[[\"depmap_id\", \"name\", \"ic50\", \"auc\"]].copy()\n",
        "train_resp.columns = [\"sample_id\", \"drug\", \"ic50_uM\", \"auc\"]\n",
        "\n",
        "# Clean ids\n",
        "train_resp[\"sample_id\"] = train_resp[\"sample_id\"].astype(str).str.strip()\n",
        "train_resp[\"drug\"] = train_resp[\"drug\"].astype(str).str.strip()\n",
        "\n",
        "# Convert IC50 to numeric\n",
        "train_resp[\"ic50_uM\"] = pd.to_numeric(train_resp[\"ic50_uM\"], errors=\"coerce\")\n",
        "\n",
        "# Convert IC50 (uM) -> ln(M)\n",
        "# IMPORTANT: if ic50_uM <= 0, log is invalid, so we drop those\n",
        "train_resp = train_resp[train_resp[\"ic50_uM\"].notna() & (train_resp[\"ic50_uM\"] > 0)].copy()\n",
        "train_resp[\"ln_ic50\"] = np.log(train_resp[\"ic50_uM\"] * 1e-6)\n",
        "\n",
        "print(\"train_resp shape:\", train_resp.shape)\n",
        "print(train_resp[[\"sample_id\", \"drug\", \"ic50_uM\", \"ln_ic50\", \"auc\"]].head())\n"
      ],
      "metadata": {
        "id": "DWikRgwxUO_U",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e7a276b0-e21b-425c-c368-e562516a17e0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "train_resp shape: (361898, 5)\n",
            "      sample_id         drug    ic50_uM    ln_ic50       auc\n",
            "533  ACH-000569   epinastine  32.524180 -10.333527  0.979333\n",
            "820  ACH-000879  floxuridine   0.501776 -14.505112  0.726373\n",
            "821  ACH-000320  floxuridine   0.007879 -18.659009  0.336463\n",
            "822  ACH-001145  floxuridine   0.041552 -16.996323  0.485618\n",
            "823  ACH-000873  floxuridine   0.097745 -16.140901  0.548483\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "CELL_EXPR_PATH = \"/content/drive/MyDrive/prisma ds/CCLE_expression.csv\"  # change this\n",
        "\n",
        "expr = pd.read_csv(CELL_EXPR_PATH)\n",
        "print(\"expr shape:\", expr.shape)\n",
        "print(expr.columns[:15])\n",
        "print(expr.head(3))\n"
      ],
      "metadata": {
        "id": "lklaQMh7a5dO",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8873882d-d0d5-405a-a0d3-7b97d558d105"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "expr shape: (1406, 19222)\n",
            "Index(['Unnamed: 0', 'TSPAN6 (7105)', 'TNMD (64102)', 'DPM1 (8813)',\n",
            "       'SCYL3 (57147)', 'C1orf112 (55732)', 'FGR (2268)', 'CFH (3075)',\n",
            "       'FUCA2 (2519)', 'GCLC (2729)', 'NFYA (4800)', 'STPG1 (90529)',\n",
            "       'NIPAL3 (57185)', 'LAS1L (81887)', 'ENPP4 (22875)'],\n",
            "      dtype='object')\n",
            "   Unnamed: 0  TSPAN6 (7105)  TNMD (64102)  DPM1 (8813)  SCYL3 (57147)  \\\n",
            "0  ACH-001113       4.331992      0.000000     7.364397       2.792855   \n",
            "1  ACH-001289       4.566815      0.584963     7.106537       2.543496   \n",
            "2  ACH-001339       3.150560      0.000000     7.379032       2.333424   \n",
            "\n",
            "   C1orf112 (55732)  FGR (2268)  CFH (3075)  FUCA2 (2519)  GCLC (2729)  ...  \\\n",
            "0          4.470537    0.028569    1.226509      3.042644     6.499686  ...   \n",
            "1          3.504620    0.000000    0.189034      3.813525     4.221104  ...   \n",
            "2          4.227279    0.056584    1.310340      6.687061     3.682573  ...   \n",
            "\n",
            "   H3C2 (8358)  H3C3 (8352)  AC098582.1 (8916)  DUS4L-BCAP29 (115253422)  \\\n",
            "0     2.689299     0.189034           0.201634                  2.130931   \n",
            "1     1.286881     1.049631           0.321928                  1.464668   \n",
            "2     0.594549     1.097611           0.831877                  2.946731   \n",
            "\n",
            "   C8orf44-SGK3 (100533105)  ELOA3B (728929)  NPBWR1 (2831)  \\\n",
            "0                  0.555816              0.0       0.275007   \n",
            "1                  0.632268              0.0       0.014355   \n",
            "2                  0.475085              0.0       0.084064   \n",
            "\n",
            "   ELOA3D (100506888)  ELOA3 (162699)  CDR1 (1038)  \n",
            "0                 0.0             0.0     0.000000  \n",
            "1                 0.0             0.0     0.000000  \n",
            "2                 0.0             0.0     0.042644  \n",
            "\n",
            "[3 rows x 19222 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "1) Clean + prepare the expression matrix (features)"
      ],
      "metadata": {
        "id": "RUn6UcGPgHsG"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "You need depmap_id to merge with PRISM.\n",
        "\n",
        "You want gene symbols only so they match DGIdb later.\n",
        "\n",
        "You force numeric values so ML won’t break."
      ],
      "metadata": {
        "id": "mTA07WzqgVyH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import re\n",
        "\n",
        "# 1) Rename the cell line id column\n",
        "expr = expr.rename(columns={\"Unnamed: 0\": \"depmap_id\"})\n",
        "\n",
        "# 2) Clean gene column names: \"TSPAN6 (7105)\" -> \"TSPAN6\"\n",
        "def clean_gene(col):\n",
        "    m = re.match(r\"^([A-Za-z0-9\\-_]+)\\s*\\(\\d+\\)$\", str(col))\n",
        "    return m.group(1) if m else str(col)\n",
        "\n",
        "expr.columns = [\"depmap_id\"] + [clean_gene(c) for c in expr.columns[1:]]\n",
        "\n",
        "# 3) Ensure numeric gene matrix (keep depmap_id as string)\n",
        "for c in expr.columns[1:]:\n",
        "    expr[c] = pd.to_numeric(expr[c], errors=\"coerce\")\n",
        "\n",
        "print(\"Expression ready:\", expr.shape)\n",
        "print(\"Example depmap_id:\", expr[\"depmap_id\"].iloc[0])\n",
        "print(\"Example genes:\", expr.columns[1:6].tolist())"
      ],
      "metadata": {
        "id": "A1XSuTbLgKxB",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f2f9ea56-1741-4d43-ce66-4f9b86a56cae"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Expression ready: (1406, 19222)\n",
            "Example depmap_id: ACH-001113\n",
            "Example genes: ['TSPAN6', 'TNMD', 'DPM1', 'SCYL3', 'C1orf112']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "2) Prepare PRISM target table (IC50/AUC per drug × cell line)"
      ],
      "metadata": {
        "id": "hYuc7i-kgZlQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Keep only what we need from PRISM\n",
        "train = prism[[cell_col, drug_col, ic50_uM_col, auc_col]].copy()\n",
        "train.columns = [\"depmap_id\", \"drug\", \"ic50_uM\", \"auc\"]\n",
        "\n",
        "# Clean types\n",
        "train[\"depmap_id\"] = train[\"depmap_id\"].astype(str).str.strip()\n",
        "train[\"drug\"] = train[\"drug\"].astype(str).str.strip()\n",
        "\n",
        "train[\"ic50_uM\"] = pd.to_numeric(train[\"ic50_uM\"], errors=\"coerce\")\n",
        "train[\"auc\"] = pd.to_numeric(train[\"auc\"], errors=\"coerce\")\n",
        "\n",
        "# Convert IC50 (uM) -> ln(IC50 in molar)  [optional but recommended]\n",
        "# molar = uM * 1e-6\n",
        "train[\"ln_ic50_M\"] = np.log(train[\"ic50_uM\"] * 1e-6)\n",
        "\n",
        "# Drop unusable rows\n",
        "train = train.replace([np.inf, -np.inf], np.nan)\n",
        "train_clean = train.dropna(subset=[\"depmap_id\", \"drug\", \"ln_ic50_M\"])\n",
        "\n",
        "print(\"PRISM clean:\", train_clean.shape)\n",
        "print(train_clean.head(3))\n"
      ],
      "metadata": {
        "id": "U0pAqBf_ghge",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "61cf9bf3-e25d-4afc-ce71-b18843e4f3af"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.12/dist-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log\n",
            "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PRISM clean: (361878, 5)\n",
            "      depmap_id         drug    ic50_uM       auc  ln_ic50_M\n",
            "533  ACH-000569   epinastine  32.524180  0.979333 -10.333527\n",
            "820  ACH-000879  floxuridine   0.501776  0.726373 -14.505112\n",
            "821  ACH-000320  floxuridine   0.007879  0.336463 -18.659009\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "3) Merge expression features with PRISM targets"
      ],
      "metadata": {
        "id": "ABtW7PY1gnS0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Merge: each row becomes (cell line + drug) with gene features\n",
        "train_full = train_clean.merge(expr, on=\"depmap_id\", how=\"inner\")\n"
      ],
      "metadata": {
        "id": "Y9FenBmygond",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 158
        },
        "outputId": "0d8d925d-f863-4aac-8eda-f43e50148d5b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'train_clean' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-378532446.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Merge: each row becomes (cell line + drug) with gene features\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtrain_full\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrain_clean\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"depmap_id\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"inner\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m: name 'train_clean' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "print(\"Merged shape:\", train_full.shape)\n",
        "print(\"Example columns:\", train_full.columns[:10].tolist())\n"
      ],
      "metadata": {
        "id": "fVMDaP1st1Tn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# 4) Build training matrix X and target y\n",
        "#Because you can’t train anything until you define X (genes) and y (IC50) and remove missingness.\n",
        "\n",
        "\n",
        "# Identify gene columns (everything except metadata)\n",
        "meta_cols = {\"depmap_id\", \"drug\", \"ic50_uM\", \"auc\", \"ln_ic50_M\"}\n",
        "gene_cols = [c for c in train_full.columns if c not in meta_cols]\n",
        "\n",
        "# X = gene expression, y = ln(IC50 in molar)\n",
        "X = train_full[gene_cols].apply(pd.to_numeric, errors=\"coerce\")\n",
        "y = pd.to_numeric(train_full[\"ln_ic50_M\"], errors=\"coerce\")\n",
        "\n",
        "# Optional: drop genes that are mostly missing\n",
        "min_non_nan = 0.95\n",
        "keep_genes = (X.notna().mean(axis=0) >= min_non_nan)\n",
        "X = X.loc[:, keep_genes]\n",
        "\n",
        "# Drop rows with too many missing gene values\n",
        "row_keep = (X.notna().mean(axis=1) >= min_non_nan) & y.notna()\n",
        "X = X.loc[row_keep].fillna(X.mean())\n",
        "y = y.loc[row_keep]\n",
        "\n",
        "# Keep metadata aligned\n",
        "drug_series = train_full.loc[row_keep, \"drug\"].astype(str).values\n",
        "depmap_series = train_full.loc[row_keep, \"depmap_id\"].astype(str).values\n",
        "\n",
        "print(\"Final X shape:\", X.shape)\n",
        "print(\"Final y shape:\", y.shape)\n",
        "print(\"Example drug:\", drug_series[0])\n"
      ],
      "metadata": {
        "id": "u811Stfbtycw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You must convert that patient sample into the SAME feature space as PRISM/DepMap\n",
        "\n",
        "Your PRISM model uses DepMap expression genes (columns like TSPAN6, TNMD, …).\n",
        "Your patient sample uses genes from your RA dataset (symbols like LBP, SEMA3A, …).\n",
        "\n",
        "So the next step is to build:\n",
        "\n",
        "✅ patient_x = one-row vector aligned to X.columns (the gene columns you trained with)"
      ],
      "metadata": {
        "id": "bH3w8km7wkKE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Cell A — Build the patient vector aligned to DepMap genes"
      ],
      "metadata": {
        "id": "MgwwnKUxwoFI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# =========================\n",
        "# Build patient vector aligned to DepMap feature genes (X.columns)\n",
        "# =========================\n",
        "\n",
        "# patient_vec: pandas Series indexed by gene symbol, values are expression for selected GSM sample\n",
        "# mat: your cohort matrix with genes as index and samples as columns (already built)\n",
        "\n",
        "# Ensure patient gene symbols match DepMap style (uppercase, remove .1 etc)\n",
        "patient_vec_clean = patient_vec.copy()\n",
        "patient_vec_clean.index = (\n",
        "    patient_vec_clean.index.astype(str)\n",
        "    .str.upper()\n",
        "    .str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
        "    .str.strip()\n",
        ")\n",
        "\n",
        "# DepMap feature genes are the columns of X (already cleaned to symbols like TSPAN6)\n",
        "depmap_genes = pd.Index([g.upper() for g in X.columns])\n",
        "\n",
        "# Align: keep only genes common between patient and DepMap features\n",
        "common_genes = depmap_genes.intersection(patient_vec_clean.index)\n",
        "\n",
        "print(\"Patient genes:\", len(patient_vec_clean))\n",
        "print(\"DepMap genes in model:\", len(depmap_genes))\n",
        "print(\"Common genes:\", len(common_genes))\n",
        "\n",
        "# Build patient vector in DepMap feature space (missing genes -> 0)\n",
        "patient_x = pd.Series(0.0, index=depmap_genes)\n",
        "patient_x.loc[common_genes] = patient_vec_clean.loc[common_genes].astype(float).values\n",
        "\n",
        "# Convert to (1, n_features)\n",
        "patient_x_arr = patient_x.values.reshape(1, -1)\n",
        "print(\"patient_x_arr shape:\", patient_x_arr.shape)\n"
      ],
      "metadata": {
        "id": "we1hY5fvwl4h"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from collections import Counter\n",
        "\n",
        "# See which drugs have enough samples\n",
        "drug_counts = Counter(drug_series)\n",
        "top_drugs = [d for d, n in drug_counts.most_common(30) if n >= 200]\n",
        "\n",
        "print(\"Top usable drugs:\", len(top_drugs))\n",
        "print(top_drugs[:10])\n"
      ],
      "metadata": {
        "id": "zzYHzjSIuFHz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import mean_squared_error, r2_score\n",
        "from xgboost import XGBRegressor\n",
        "import numpy as np\n",
        "\n",
        "# Pick one drug to test\n",
        "DRUG_TO_TEST = top_drugs[0]\n",
        "\n",
        "mask = (drug_series == DRUG_TO_TEST)\n",
        "X_d = X.loc[mask]\n",
        "y_d = y.loc[mask]\n",
        "\n",
        "X_train, X_test, y_train, y_test = train_test_split(\n",
        "    X_d, y_d, test_size=0.2, random_state=42\n",
        ")\n",
        "\n",
        "model = XGBRegressor(\n",
        "    n_estimators=400,\n",
        "    max_depth=6,\n",
        "    learning_rate=0.05,\n",
        "    subsample=0.9,\n",
        "    colsample_bytree=0.6,\n",
        "    reg_lambda=1.0,\n",
        "    random_state=42\n",
        ")\n",
        "\n",
        "model.fit(X_train, y_train)\n",
        "pred = model.predict(X_test)\n",
        "\n",
        "rmse = np.sqrt(mean_squared_error(y_test, pred))\n",
        "r2 = r2_score(y_test, pred)\n",
        "\n",
        "print(\"Drug:\", DRUG_TO_TEST)\n",
        "print(\"Samples:\", X_d.shape[0])\n",
        "print(\"RMSE:\", rmse)\n",
        "print(\"R2:\", r2)\n"
      ],
      "metadata": {
        "id": "79ik--X3uHto"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Cell B — Predict IC50 for ONE drug (the same one you trained)"
      ],
      "metadata": {
        "id": "nphcZfACwuYl"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "This gives you your first “digital twin” output:\n",
        "\n",
        "For this drug, predicted sensitivity for this patient state."
      ],
      "metadata": {
        "id": "U1wiFqQ8w9wc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# =========================\n",
        "# Predict IC50 for the selected patient sample for this drug\n",
        "# =========================\n",
        "pred_ln_ic50 = float(model.predict(patient_x_arr)[0])\n",
        "\n",
        "# Convert back to IC50 in microM (reverse ln(ic50_uM*1e-6))\n",
        "pred_ic50_M = np.exp(pred_ln_ic50)\n",
        "pred_ic50_uM = pred_ic50_M * 1e6\n",
        "\n",
        "print(\"Patient sample:\", patient_sample)\n",
        "print(\"Drug:\", DRUG_TO_TEST)\n",
        "print(\"Predicted ln(IC50_M):\", pred_ln_ic50)\n",
        "print(\"Predicted IC50 (uM):\", pred_ic50_uM)\n"
      ],
      "metadata": {
        "id": "Ndizcsyfw69D"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Cell C — Predict across many drugs (rank safest/most effective candidates)"
      ],
      "metadata": {
        "id": "RjhKjIOExBRb"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "This is the real useful output: rank drugs by predicted IC50."
      ],
      "metadata": {
        "id": "rVVt8F-wxDPY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# =========================\n",
        "# Train-per-drug quick ranking (MVP-friendly)\n",
        "# =========================\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import mean_squared_error\n",
        "from xgboost import XGBRegressor\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "results = []\n",
        "\n",
        "# Choose drugs with enough training samples\n",
        "from collections import Counter\n",
        "drug_counts = Counter(drug_series)\n",
        "usable_drugs = [d for d, n in drug_counts.items() if n >= 300]  # adjust threshold if needed\n",
        "\n",
        "print(\"Usable drugs:\", len(usable_drugs))\n",
        "\n",
        "for drug in usable_drugs[:30]:  # start with first 30 to keep runtime reasonable\n",
        "    mask = (drug_series == drug)\n",
        "    X_d = X.loc[mask]\n",
        "    y_d = y.loc[mask]\n",
        "\n",
        "    if len(y_d) < 300:\n",
        "        continue\n",
        "\n",
        "    X_train, X_test, y_train, y_test = train_test_split(\n",
        "        X_d, y_d, test_size=0.2, random_state=42\n",
        "    )\n",
        "\n",
        "    m = XGBRegressor(\n",
        "        n_estimators=300,\n",
        "        max_depth=6,\n",
        "        learning_rate=0.05,\n",
        "        subsample=0.9,\n",
        "        colsample_bytree=0.6,\n",
        "        reg_lambda=1.0,\n",
        "        random_state=42\n",
        "    )\n",
        "    m.fit(X_train, y_train)\n",
        "\n",
        "    # evaluate quickly\n",
        "    pred_test = m.predict(X_test)\n",
        "    rmse = float(np.sqrt(mean_squared_error(y_test, pred_test)))\n",
        "\n",
        "    # predict patient\n",
        "    pred_ln = float(m.predict(patient_x_arr)[0])\n",
        "    pred_uM = float(np.exp(pred_ln) * 1e6)\n",
        "\n",
        "    results.append({\n",
        "        \"drug\": drug,\n",
        "        \"n_samples\": int(len(y_d)),\n",
        "        \"rmse\": rmse,\n",
        "        \"pred_ic50_uM_for_patient\": pred_uM\n",
        "    })\n",
        "\n",
        "rank_df = pd.DataFrame(results).sort_values(\"pred_ic50_uM_for_patient\", ascending=True)\n",
        "rank_df.head(15)\n"
      ],
      "metadata": {
        "id": "8Wu8Gdk0xE_n"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}